{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generating the toy dataset to be phylogenetically curated as part of the course in [this]() notebook.\n",
    "\n",
    "For demonstration purposes we are extracting a random subset of sequencing records (max 5 per species) and a set of records that we know are mislabelled from the full-scale curation process [here](https://github.com/HullUni-bioinformatics/metabarcode-course-2016/tree/master/data/exercise-5/supplementary_data/reference_db)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from reprophylo import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:Cloud:Log file (/root/.picloud/cloud.log) opened\n"
     ]
    }
   ],
   "source": [
    "pj = Project('../../input_data/target_locus.csv', pickle='reduced_tree.pkl', git=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genbank formatted sequence records from full analyses after removal of redundancy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/reprophylo/reprophylo.py:1015: UserWarning: Version control off\n",
      "  warnings.warn('Version control off')\n"
     ]
    }
   ],
   "source": [
    "pj.read_embl_genbank(['../../input_data/CytB_nr_pre_Sativa.gb'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.extract_by_locus()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Distribution Of Sequence Lengths\n",
      "Distribution Of Sequence Statistic \"Gc_Content\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/pymodules/python2.7/matplotlib/figure.py:371: UserWarning: matplotlib is currently using a non-GUI backend, so cannot show the figure\n",
      "  \"matplotlib is currently using a non-GUI backend, \"\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAFEAAAInCAYAAADga7XDAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFKdJREFUeJztnXmUHNV9Ru8gUJCEJCyEEDEQAUZIGAyCmCWYoABmix3A\nG0tCwhKOYgzYTmIw+Jx4YswSMAlLbCNMEOAYAkHGhoMgLDEQE2yxDQzLYASIRSABwsgsQgLN5I9f\ndbqm1d0z1fNNd9X0d8/p09VV1dVv7tTyqt6rr8AYY4wxxgi5AlgGdKfGnQk8CnQBdwGbp6adDjwD\n9AD7p8bvkizjGeCiYSxvLtkLmEV/ieNTwycDlyfD2xFi1wOmAYuAjmTaQmDXZHgBcODwFLc26zT7\nB1P8D/DbinFvp4Y3AN5Ihg8BrgU+ABYTEncDNiXEL0zmuxo4dHiKW5t1m/2Dg+As4GhgJeU17PeB\nX6XmeRn4KCH15dT4Jcn4ptLKNbEW3wK2AOYBF7a4LIMijxJLXAN8MhleQv+DzGbEGrgkGU6PX1Jt\nYePGjVsE9A3x1TX0P0vPNPofWLZJDZ8M/DgZLh1YRgNbAs9SPrD8mtg/dlD/wNI3QFk6B1HegZbR\ndK4FXgFWAy8BxwE3EFK7gPnAlNT8ZxAHlB7ggNT4UhVnEXBxnd8bkRKbzbBJzPM+sdnc3eoCFAHF\npug1cbiwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBVii\nAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFFAHu93bjqd\nnZ19qeGOevNWw2uiAEtM0chaCJYowRIFWKIASxRgiQIsUYAlCrBEAa2UWC217nzgKSK57qfAxNS0\n3KbWtVLiPNYOArod+DiwI/AbQhxEuNDhyfuBwA8ohwv9EDieCCbapsoyh528pdbdAfQmw7+mHGGV\n69S6PO8TjyNiqyBS69LpdKXUusrxTq1L8S0i/uqaVhdkMOTxeuIxwMHAvqlxQ06tA9h1113np4af\nnDNnzlMA8+fH6O7u7qPS88+dO3fmwoULtyt9XrhwIXlkGv2PzgcCTwCTK+Yb1tS6zs7OvvSF2TpU\nnaeVa+K1wN6EsJeAbxNH49HEAQbgfuBE4Eng+uT9w2Rc6Q86EbgSGENIvK0ppU/RSolHVhl3RZ35\nz05elTwE7CApUYPk9cBSKCxRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJ\nAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJKcZHV2dTh5p9sifBsj54\nelLc4ZV5GV4TBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFFAQ6GLBaWPKn9vX19fvysz\nHR0d9ZxUXYbXRAF5DNJoKh0dHR2TYNlyeGsj2LCRZXhNFGCJAixRQN5S675INBatAXaumD+3qXWt\nZC9gFv0lzgCmA7+gv8RSuNB6RCDRIspVjYXArslwQ+FCRW6oqpZa10NE/lXi1DoBTq0b6RSlsj2s\nqXWf3mWX9deMHj1+9OrV63c/9ND/J9cVNbWuxC+Io26JYU2tG+qBpZVcC7xChE2+RCR3HpoMrwSW\nArem5j+DOKD0AAekxpeqOIuAi+v83oiU2GxGZBVnxGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixR\ngCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBL\nFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKaCQX5yPA\nFCJU4nXWjqpqOwazJo4CPkc5gmU58BQRrbKcCPO5BjhskMsrUS21bhJwB5EXdjv9kzULmVq3LnAy\nIakXeJsI/bkc+CfgvGT4buCdZJ4lwEmE+IGollp3HnBqMnwacG4ynOvUuno8DbwL/Bswm/piRgH7\nAPOS7/QM8jem0V9iD7BJMjw1tZzTCaklbgN2J1LrnkqNPwK4tMZvtSQXZwGwFXA8sbatqTPvGuC/\ngWOBrYk/shE2ofyHLKMsNNepdfUOLF9vcJlLga81+N00fRQkVipvqXXLiM14KbGpvpaMH3GpdbsR\nO/zrktc5ybhGmMbaB5bSvu+brH1gKXxq3SiiWtJb43Ulgzsql6hMrTuWqOLcSfUqzohIrfs2Ieun\nxJFxYvLaA7gxmdY51B8ZRnIh8QVi7ahGB1FJfmGoPzKM5CL6bwrw8zoL/znlKklbkUXiM8SRsxZT\niQq6qcORxMWGnapMm5VMO6KpJcrGsG3OWeqJ04HngAeI/V/pdGs7YD/gsWSef6j43ncy/MaIp1bV\nZqBXXsjFmrhVhnnbiiwSFw9XIYpOo+fOY4E/SIZfAN7TFKeYZG1j+TgRIL6CeG7KE8BbybjttUUr\nDlnWxFnAPcA41j467w/sCewNPKIs4EjjTmKtq3xiD8m4Fck8eSUX587vAN+tM/0somkgr+Ti3Pl9\n4NU6018lHrzQdmSRuAD4szrTP5vM03Zkkfi3wEbADUQT5fjktRswn7ig2mi7zIill2jF660yXOtV\nr0Ww1bTktO/qwZRsMD8y0qkn8ZhmFaLouFeYgHoSPzKE5Q7lu4WjnsTFRAvfRhmWNwU4E3h+CGUa\nUXyZ6IGwimgSPQHYEdggNc944pTvK8AtxGMylwJzmlrSwdGy076JRI+sxfSvyqxKXulxi4BvEGLz\nyLBJ7Kg2sgrrEBXsvYmrNhtT7in7ONFv8aFBLqtV9FHj700/rv3N+s2+VZcx2EthvcCvkpepwFUc\nAZYowBIFWKIASxRgiQIsUUDWxvt1iM5LHyPOqatVXt2BqQ7bAE9SnA5MleSiQ9MlRKemU4nTvOUZ\nvmsS3iXu6SsquWh3XkV08jQVZJH4X0R/G1NB1nbnPYC/J+5sMgn1ric+z9r7gA2AyUT78iv0b2fu\nSObPa4/allxPbOTGHrc7VzC7WYUoOln2iX9MNAvUYuNknrYji8S7gU/Xmb4vUQlvO5QXIEbRpvtE\npcQ9gDeEyysMA507f5XIcyitYRdSvcvxJGACcVN52zGQxBWUqzrTiDXttYp5+ohbMe4H/kVZuJHI\nYuCQVhdiCOTiUti0DPO2FW4eEJBFYqlPdr3XO8TV74uI9KS2IIvEq4nolA4ipuCm5PV0Mq6biLda\nQwS1dZHfixEtYx/itrSDqkw7mDiS75t8/lMi7+YnDf7WV4l/yuPJMDQWC5gmF7el3Q+cX2f6+fTv\nNXYp9e/AqsX2hMD1ibOgO4gQtyyxgNW2sFw0D+xI/eaB54FPpD53ka2rcokZRHTV+8Su4R7g88Td\nXFcl81wFHJoMH0KkPX1AVMMWUc5TbApZJK4g2pxrsS/wu9TnCcl3svI4EVA5ibg5/WAiSC1rLGDT\nyFJPvJY4BZwL/DPxH+8jEki+TsShpuNI/4Q4Umelh2hVvJ1oYexi7Tu1BooFrDotD6l1Y4nO7aVG\n+g+SV+nzrck8EPuzc6h+EMrKWUQn/B7K4UabUk75/GbyKnEb1VP0cnFgKXEw8EOisLcRB5CDVQtP\nmJK8b0Hc4T+RxmIB0+RKYjO4l7io0UXsFqCxWMA0bSdxOMjFBQiIGJc51O8Vtk/GZRaeLBIPAn5G\nVGrfAd6sMk/br+4D0UWE3v5hqwvSILk4Y5lBNA88mOE7bUEWiW8QPcNMBVkvhX1+uApSZLIcWK4k\n6mw3Ead3z1E9OOPFoRerWGSRmH4ow2dqzNNHtqztEUEWiYO5K8BVnBFOLqo4pgZZJU4gwjXuI9o0\n9kjGTyZSjWfoilYcsuwTNybklS43bQ2MSaYtB/6KiHJpu7ywLBK/S1yS353on5Puk9NHVH3a7uID\nZNucP0NcjK0VmPEccRG17cgicTKxH6xFL9Es0HZkkbiM2A/WYifa8GwFskm8hXhyWrU+NrsBf0nt\nR42YhE2J64lLgcuIzfcq4D+ILiOLaayxvlnkpo1lC2Jtq0z4vJn+Ty3LI7lpY3mR6LYxEdiWaGNZ\nRJvf+9zoswdWEM8QNfjcWUK9NbGXOndmVlCaz9cTK3C68SBxurEA7xMFWKIASxRgiQIsUYAlAqd0\ndk75x87O6Y1+3xJTfNDgTe9Zzp0bDcm4t8HvNZ2345F6mcki8e7kvdqpYK3Tw7Y4Dcwi8Xjixsdt\niHv20s/tO4rokH6JtHQFIYvECUTb8wziCneaM4l7/yYAF2uKVhyyHFhOIu6mqhQIcSvY3GSetiOL\nxM2p/3DD93C784AsBo6metvymGTa4qEXqXhk2SdeQGyyDwLfp9zpcybxUJuZwN9IS1cQskj8EfFE\n3bMIiWlWEuGUl4nKVSiyNlRdSPTd3p9yvsOzxN3xb+mKVSwaae17C7heXZAi04jELYk78KcQle7F\nxG2yU4nG77a71yXrBYjziJ5hc4mO8KVNegxxBnOirmjFIYvEOcTB41+JfWL6XHkF0b2k1q0ZI5os\nm/OJxF2mXyP6KlbSTTxNre3IsiZOJ+54r8XrVJc74sn6uPZxdaZvQZtWc7JIfICIbanG+sRp331D\nLlEBySLxPOCPgH+nnMS0KXAgkaK0OfA9aekKQpYDy53EufHFxEVYgB8n76uAvwb+V1e04pC1sn0Z\n0Sv2C8QFhw7iivb1VL/O2CinA39B9EzrBo4l9sfXEWEei4EvUd4Hnw4cR/TaPYX6B0A5jZyxvMrw\nNgNMIx6DPJNYw68DjiAake6gHDJUSmbaDjg8ef8oscVMp4mPehpKk+l6xD7yizTYSlaD3xHxWWOJ\nf/JY4okbhU2tm03sAysfr7ElcWfVL4k15TFgnqhMbxLXLl8k5L1FrIGFTa07hljbTqkYfyURFnkf\n0Xd7f+IGyXuSaUNha+KsaBpxOvmfxP4xjTS1bv78GN3d3X1Uen5Vat2TxLlymhnE/uae1LgxRLbs\nXQ39Sn8OBy5PfT6auAj8FMOUWtfZ2dnX2dk5mF6+Dd00PpU4+qaZnbyn/9CVwDX0T/JslB7iTtYx\nxNF/P+KfeTOxtpO8/ywZvok48JRS67ahyXc2DLQ5/x4hKE1pp31PxfiX6J8k1yiPEv3FHyTW+IeJ\nqtV4oip1POUqDoTg65P3D4kLJU3tOz6QxJdY+8j7KeJe58qbIceiO3c+L3mleZPacaxnJ6+WMNDm\nfC9x4+MOyefDiMS6W6vMuz3aCndhGEjiucQm3UWsffOJ+tgFFfONIupxv1QXsAgMJPE54kLrAmJz\nWpB8frxivn2S6W15q+5gTvseBD47wDx3EJtzW+KesgIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAl\nCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRg\niQIsUYAlCrBEAZYowBIFWKIASxRgiQIsUYAlCrBEAZYowBIFWKIASxRgiQLyKHFb4JHUawWREDWJ\nuMP/N0QqXTo+5nQidbmHSItqKnmU+DQwK3ntQjxd40YihekOIpHuLsqpTOnUugOBH9DkvyuPEtPs\nR6TQvUSBU+tazRGEIChwal0rGU2koJxWZVqhUutaySFEAl2JHgqaWtdKjqS8KUOk0xUyta5VjCMO\nKiekxp1LQVPrWsW7rB3BX9jUOjMILFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBVii\nAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGW\nKMASBViiAEsUYIkCLFGAJQqwRAGWKMASBViiAEsUYIkCLFGAJQqwRAGWKCCvEjcEbgCeIqJadsOp\ndZm5CFgAzAQ+Qchxal0GJgJ7AVcknz8kMhSdWpeBLYHXgXnAw8CPiMQmp9ZlYF1gZ+Ak4AHgQvoH\nqoFT6wZkKvB86vOngFuIg4xT6wbJUiK5c3ryeT/gCeBmnFqXiZOBnxBingWOBUbh1LpMPAp8ssp4\np9aNVCxRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijA\nEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUKsEQBlijAEgVYogBLFGCJAixRgCUK\nsEQBlijAEgVYogBLFGCJAixRgCUKsEQBliggrxIXA48Bj1DOuHFqXUb6gNnALMrhaU6ta4COis9O\nrctIH3An8CBwQjJu2FPrBpkXVhg2Td43BrqIPMXfVszzZvJ+CfDnqfGXA5+rsswuyml3jb66qhU2\nrzFXrybvrwM3EpvnMiK1bikh+bVkniXA5qnvbpaMq2SnYSkp+dycxwLjk+FxxNG2m0iny2VqXR7Z\nkthsuoDHieoLRBXnTqpXcc4gDig9wAFNK6kxakYDG7S6EEXiS8A5lOuZZwKrgV5id7FRi8pVGP6O\nEPYA8A5wAVETOBX4BlHHvLRlpSsITwPHJMN7EmvfF1LTDwJebHKZCsf7lOuUHcAqYNvU9M2I08dM\n5LGeOJyMJsRBnIGspr+0D4kA4Ezk9YxlONkRWE6siesA21Ouc05uZIGVV0pGOr2DnC/TFtpua+JW\nrS7ASGILqq9tHck0Mwh6gSlVxk8G1mRdWLsdnQdiHFENykS77RMvSQ2fDbyX+rwucd3y0awLbTeJ\nO6SGZxL1xBKrgYeA7zW1RAVmHjBBtbB23Sc+BqzX6kIUnReJ07/r8ZXwhlmHkHcdsBJ4AfgOMK2F\nZSo0k4gnET1MXIC4EziSBi5EtDt7Es8JXEU0eC0nnpNV60lEJmEqcBrRQriSeCbW7GTaGOAsYjM3\nNbiZuI7YDZxCbNKVTGHwV33akiuA3QeYpwMfaKpyENFzbGKVaRsSm+9hzSxQEVkAfKXO9C8TT6A0\ndXiZ/g1TlcygemeourTbad9k6h8s+mig8b7dJL5MNFTVYgcaWBPbjYuIZ56OrTJtHPHk3ouzLrTd\nWvs2IU7veoHvE9IgOs2fRPjYmehIauowjThKryFk9ibDC4i+kZlptzUxzSTgY4SDZyj3ATfGGGNM\n4fk/hY2hBxypXewAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe807cd2050>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAEUAAAInCAYAAADOLpRRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFMZJREFUeJztnXuYHWV9xz9ZcBOyAZLNhVAjRDCBoCCCBpAiEBOUS7m0\nhVa0ELFKgQKtXGukriIXIVZtipWqAUr1eYKk+pSHa4KkFCuEWySBJBBI5JaEQILcwm7I2f7xm9Mz\ne/I9Z86cy8yc2d/nec6z58yZmX33s3N5533f3+8Fx3Ecx3EywJC0CwCsAd4AtgJbgKnAtcBxQB/w\nLPBF4A8plS8VVgPdZctmAB3B+6uDV2J0RK+SCOVH7AKgELx/CJiQZGGyIKUfWAg8AnxZfH8GcEei\nJcoAuwY/xwJLgMNC380C5ideoozxDeCC4P1M4DfAMLViV1fXKuwoa+S1pDV/RmMMB3YM3ndhEo4C\nPgs8CYypsm1/xL57avj9ch/b17BhK9kF+GXwfnvgZ8A9wDNAJ3bBBfgtcHZShUpbympgf7F8UtIF\nCZOFu0+rWJR2AdIg6ppS9z7yfKTUjUsRpH2hbQk9PT39ofexH3r9SBHkWko9RwnkXEq9uBSBSxG4\nFIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSB\nSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxFkYcT1GrYNwe0G5gG7B9+fAryeVIGycKT0\nA0cAH8OEAFyKBUBNBu4NPidGFqTAtiG4xwM3Be9vAk5MsjBZkKJCcHcB1gfv1wefEyML15RDgbVY\nCO4CYEXZ98XI0MTIgpS1wc8NWJDlVOzoGA+sw+KWX1EbTp06dX7o/VNnnnnmcoD5823x0qVLTw2v\nf/31109ZvHjxPsXPixcvbtof0UwqheBeA1wSLL8Uneug4tHT09PTH475qUJbheA+AtwCfInSLTkx\n0pZSKQR3IzA94bL8P1m4+2QOlyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInAp\nApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKXInApApcicCkClyJwKQKX\nInApApcicCkClyJwKQKXInApgixI2Q54HLgt+DwVWBwsexj4RNIFyoKU84GnKIWuXQNchsUp/2Pw\nOVHSljIBOAb4CaXY5LXAzsH7kcBLSRcq7XC57wEXATuFll0KPADMxv5phyRdqDSlHIeF1j6OhfUX\n+SlwHhZweTIwF5tpexvyGIJ7JfACFly5FngbuBlLBlFkCJXnXm9ZCG6a15SvAR8APgj8JfBr4K+A\nVcDhwTrTgKeTLlja15Qwxf/aV4DrgKHA5uBzomRFyn8HL7BA7YNSLEvqt+RM4lIELkXgUgQuReBS\nBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQu\nReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkWQayndsH5HeDLtctRDeQgu\nwLnAcmAZ8J0K21WM/OqG9f2wsh9WdpfmSK15H1mI9ymG4BZnrjwSmwV3P2z+5LFJFyjt00eF4J4F\nXIUJAZvzNFHSllIMwS2Elk0CPgU8CCwCPp50obIYgrs9MAo4GAvpvwXYQ+2gUgjujAMPHLa1s3NH\ngM6+vmFLH330VGjvENw7KUWbgkWfjhbbt+xCmxUOp3T3ORP4ZvB+MvB8hW1yffcpUizg3OC1FOgD\nTku6IOUTwrcT/VQofzesfw1eBxgNIzdWnsNd7iPtu08mcSkClyJwKQKXInApApcicCkClyJwKQKX\nInApApcicCkClyJwKYK4LW/Dgb2AcVgDzQZgJfBOk8uVKrVI6QZmYsktDxTbbAEeBX4B3Ahsal7x\n0qFac+RILFnu2VjKwpVYX8yzwGvBtt3Ah7DuiMnAu8APgcupnAizWbSsObIar2IJci/Dkl5GsSfw\nDeDlYNtWk0oXx/nYERKXYcDf1bFdXFomxVvzvTW/NhqRcjJwFzYo5h4swW4uqFfKOcBNWD3lv7Ce\nvBuBS5pTrGxTSdoyLAVzmO8Ba1pamoG07EIbdaQ8hlXYyukCXixb9hIwImJ/bUGUlPuB/wW+C+wQ\nWn4bdrqcgaVvvwD4OgPHreWag4AngOeAzwTLdgD+FXvmKWA12RsojVtLgtTHp7wPmIVJ+A9gTLC8\nA3s43K6RnddJateUIluAK4D9scF7K7BxIwVsiNbWGveTW4YAfw1sxOontTwXtYJUj5Q/xU6Z/wQu\nxE6lnwD7YFXpZdgIx0FTOz4PO0UWYROGvAPcUbbOn2AD+h4DDkiwbKldaJ8F/iX0+VhMUvkQzh2B\nOUBvPb+kTlI7fcZgp0eR4uD/7rL13sTG0x9ODohqjvwt8PfY2PlNWIPTG9jdR/Fg84qWXfbEKm6F\n4PUqcEKqJSqR2jjaZ7G6ySSsFrsCq73mmlpa8wtYo/WgodqFdlQD+21k29SpJmUN1jqvggUqMQ7r\n3ljdQJkyzVnYc00vVnH7MvBRBraZ7IhV2M4BbseekdZhQQetJrXK287AP2BHTSH06g1e4WWrsOp+\nUs0HqXdxdGBzjh6OPfOMpdSXvAy4D+s6TZKWdXHU2sFewCpmg6JylgVUCC5YE2eBbR8pishDv19Q\n5XdnbnLGIuWz4IJN2jgD+H0qJUqZCcBCLBY5fKT8AotLXk3MIwWSa45sFSoE9wSs++SJVEpE9kJw\nh2NznoZnva14hxwsIbi3Yof76uC1BasjjRPbp97FAdavU23O0alYlGg9hENww2T+mnI61r5SiT2w\nsXH1ogqYSjB1M68pXZSSNsQlPAtuGBnO32qipOwevIoXuylYcoZyRmMPkKuaV7Ts0sPAh75qr/dI\nNto8tebIX1EaczIX+De2ff7pB94CFmN3k7YnSsqS4AUwEZiP5SDINXEutD2tKkTWiHv36QCmY6Os\nR6Nrm99qtFDtxCTsaTbqgpsUmcifMgerN1yMtbS9FmPbtiKOlMOAHwCzW1SWzBCnmt+LjXvLPXGk\n3A0c2qqCZIk4Ur4KHIKNZupsTXGyQZwojtXYQ98YbODfywwcADgEu5on9RCXehcHWCNyVCRVpvOm\n1UocKUe0qhBZI+2G60xSj5TDsYHGPwb2DpaNwNpZ2noIRj1shyW3LFbntwLTgu92wAYbz0qwPJlo\no70EG2j8VawFLnzB3YwN1zg6xv4ySxwpp2FZQL+Pfu5ZgT09tz1xpEzEYn8q8To5uabEkfImlftg\nwLo/Ek+93AriSHkA+EKFbUZhUWL3NaNQaRNHyhVYPoNfY/3AYGNs/wbrDx4BXN3U0rUJx2ID/cpb\n29YBRyVclky0vIGNgJyIjQoo3pafxpoVcpVDpVZ2w4ZKVGJ4sE5SZKLytgY4scr3x5OTQcXNfCDM\nzcNlM/+QvQkadtqdqAvt6cGryCws0rSc0cBHsOeftidKyigGNi+OxZokwxQ72H9Ksk/JmaAAfD7t\nQoTIRD0lNxfSKAbNHxqHuFI+hzUfbGBgNX9r6GfbE+f0uQibv+tVbDSTamgadF0c5wAPYe2ym1tT\nnGwQ5/QZjzVH5loIxJPyLJZPMvOc19Mz7ps9PZPr3T6OlNlYbTbJFEMNMRpGbqkjj2Wca0oBqwgt\nx8bpP4e+2/x73EK0iiod61WJI+WG0PtK1fl+MiSlXuJImRa9Sj6II2VRqwqRNRqp5o+hlNosV8SV\n8n7smvEHLNTtFSzZzE3Bd/VSHoZ7LXZB/x2WIGvnBvYdmzhSdgMewTrEngN+HrxWY2lXH8ZCZ+uh\nPAz3HuDDWG6Fp7HUAokRR8rlWOXtOOBjmJwvYAkgjsW6VL9dRxnUTLgLKI3efihYJzHiSDkKyxdZ\nntIMbD7SH1LKLRkHFYYb5owKv7NlxJEyCjuUK7GK+KMOwmG4aoDhLCwB8M9j7rch4tySX8IizX9U\n4fvD2DZHbRSfxPqLjsGyqO+EXchPw4I0jwE+XWnjSnHJ8+fb4qVLl54aXr8VcclXYYf41Qy8G+wc\n+u7KBvYfDsP9LJZTrtotv2LbTU9PT39PT08tbTsNt9F+GzsaLsYyVrwcLH8/dhr+hvoutEWKg5PB\nIkY6sQsuWL65sxvYdyziSHkbO31mAidRyi56N9bfcyMWXFkviyjVmic1sJ+GiTvqYAs2VPTHLShL\nZvDWfIFLEbgUgUsRuBSBSxHUIuUs4JSIdf6CZFIjJkKUlJOA64geobQJe4LOSgLfhoiS8nms3/ie\niPXuwar5p0es1xZESTmI2tsy7sLyMrU9UVLGUXtzwMvoLFttR5SUd7A2jlrYiZyMuo6Ssorac+Ef\nBjzTWHGyQZSU27A7yicj1jsYG42di0mPoqTMwYZy3Q58hW0naxyGpXm+E2trndPsAqZBlJRN2JHS\nh7XNvo7laLofa2zeBFwffH8COZjsFWprZFqMdUpdBPwZlhK1yPNY8qprqD5eta2oteVtHdYuewE2\naGcnbE6ON1tUrlSpJ03im+RURpFapOwa/Fwb/ByGjZQs7x54Acs43PZESdkL63+ZhY2hBQugvFas\n+x52EW77ukrU3eeLWA6D74vvLsRGN03DevHeAr7U1NKlRNSRMg2b0FVNe7WEgaOb5pGTIWBRR8ok\nrD5SCytIuROrWUQdKV3YaRFmI1ZXKU9v9gbbBki1JVFSXqd09ylSYODkakXGY8O+2p6o02cZtUem\nzyAnaVmjpNyKJaiKans9MVjv1saLlD5RUuZiF9B52Ji33cu+n4gNv5iHjWasN0V8poi6pryLDcG6\nHavAfQ27oL6BPf8UB++sCNbLxcxztfT7PIeNgDwfy6FSAP4o+Pk/2PynB5CTkH6o/YFwM9aAlItG\npCi821TgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkWQBSnl4bfdWJzP09gAw8QzcWRBSnn4\n7aWYlMnAvcHnRElbigq/PR4L/ib4WS1fZUtIW4oKv92F0liX9dSZ7qMR0pQSFX4LdkolnvwqzamB\nVfjtzdjRMR4bKLQrJk6ShRDcVhIOv70Gm7gA7CJbKW92y0Jw076mhCkW8Gqst/FpbBRD4snE0zx9\nwoRnwd2IzXWYGlk6UjKDSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsR\nuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgUgUsRuBSBSxG4FIFLEbgU\ngUsRuBSBSxG4FIFLEbgUQdpShmETui7BwnCvCpZPxRKOP47NrvuJJAuVdhDUu9jchu8EZXkA+GMs\n++Bl2ByHR2MhdEcmVai0jxQozd/RiUWzb8KCKovZBkdi86omRtpHCtg/5jFgT2zipCexgMoHgNnB\n94ckWaAsSCkA+2NHxt1YttKvY+kXfwmcjCXqnFG+Yd5DcItchiUUfyO0bAg6+W9uQ3DHUMp6sQN2\nNCxh4BQ606g+eXXTSfv02RVL8tARvG4GFmIzxFyHzRKzOficGGlLWYqlbS3nEWy+slRI+/TJJC5F\n4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgS5llJj4/U25FrKYKQZeWoz2cWR\nSVyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhcisClCFyKwKUIXIrApQhc\nisClCFyKIE0plcJvAc4FlgPLgO8kX7R0GR783B54EAu/PRKbGvh9wXdjK2zbsr7krDAcC8r+MDAP\nC5GLIrcd7B3Y6bMeuA8Lv50MfAo7chYBH0+6UGmHy6nw2+2BUcDBWDj/LcAeauNKIbiVaMcQ3GL4\n7Z2UIk3BIk9Hi/Vzefqo8NvHgV9RuqZMxsL9X0u8dCmxLxbOvwR4Apt2HOyuczMWifoodkopcn/3\nqYdcnj6ZxaUIXIrApQhcisClCFyKwKUIXIrApQjSbjpoOvsNmzB5n6Hjp3d1DO18u9Db91TvuoVP\nvPtirJxOuZKy37AJk2eMmDLzwrEz+orLZm9YMBO4MY6YXJ0++wwdPz0sBODCsTP6pgwdPz3OfnIl\npatjaKdaPqLC8krkSsrbhd4+tfytCssrkSspT/WuWzh7w4IBR8XsDQs6l/euWxhnP0OaW6xE6UeU\nf79hEyZPGTp++oiOoZ1vFXr7lle/+8h9tDPe8pYkLkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F\n4FIELkXgUgQuReBSBC5F4FIELkXgUgQuReBSBC5F4FIELkXgUgQuRZC2lGphuAAXYCF13XXs+4iG\nSpYyKgwX4APAXcBqtJSoUQc9NfzuzI46KJ8Fd2Pw+Z+Ai9MoUBaklIfhPgWcALyIhdENanbGTp9j\ngp87BctXo6NNl2CHfyOvJaogWRvadBlW2HMpnVYTsEmkpwKvpFSuRCkPw70f+HTZOpUutC0j7RHX\nahbce8vWyV2YrJMhOoERaRciTU7BasInBp8vB/qwmvBC9J0r11yACXgYeAv4LrAWq/RdhNV1fpRa\n6VJiJTAzeH8odnT8eej7o4HnEy5T6ryLPSOB1bl6gb1C308AtsTdaRaq+Y3QiYkAu3X3MVDCe9jz\nVCzSrqc0g49i+VWGYP/kj1CqEI6pZ4dZq+bHpVDjerHOiHY/UmSuJsfYDX00DAm+G5QUgHFi+Rhg\na9ydtfvdJ4ou7LYdi3a/pswJvb+SUhsM2N82Ffhd3J22u5R9Q++nYPWUIn1Y9q/ZiZYoQ9xAqfmy\nYfJyTXmCUlZSJ+B5rLp/C/CZlMuSGTowGfOAzcDvgW8BE1MsU6boxnoDHsMeCBcCn6OOB8O8cSgw\nFzulVmEPjC8AsbJj5IHxwCXACuwU+hmlTvYdgCuw02rQcBvWjrIUOA/dTzSO2p+qc8FcLNNxNYYw\nSC68RwNrsH7ockZip8tJSRYoC9wBnFPl+7OwMS6DihcZ2FBdzt5Y53ws2r2aP4bqF89+6ugMa3cp\nL2IN15XYlzqOlHbnB9jIp+Hiuy5sPo9/jrvTdm/N3wWrzheA6zAJAPsAf4v9fQcA61IpXYpMxO5C\nWzE5heD9HcAH69lhux8pYbqBD2F/0zOURlk6juM4Thvwfx0rOQgoRxJ2AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fe807bd3f10>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "pj.report_seq_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect all record ids in a list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1760\n"
     ]
    }
   ],
   "source": [
    "full_id_list = []\n",
    "for r in pj.records:\n",
    "    full_id_list.append(r.id)\n",
    "    \n",
    "print len(full_id_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Find the ids of the records which are mislabelled according to SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "rec_mislabelled = {}\n",
    "sativa=open('../../input_data/CytB.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'EU492281.1': '0.989', 'EU224046.1': '0.838', 'AJ969128.1': '0.999', 'AJ937943.1': '0.868', 'JF489783.1': '1.000', 'KT275288.1': '0.994', 'DQ185405.1': '1.000', 'KT275289.1': '0.998', 'EU224045.1': '0.819', 'JQ346141.1': '1.000', 'JQ661401.1': '1.000', 'AJ937931.1': '0.922', 'AJ937952.1': '0.791', 'AJ937925.1': '0.999', 'JQ661398.1': '0.973', 'AJ937951.1': '0.792', 'KP644340.1': '1.000', 'DQ664351.1': '0.587', 'GQ279764.1': '0.625', 'GU182336.1': '0.920', 'JQ661399.1': '0.974', 'KP452507.1': '0.508', 'JQ661400.1': '0.973', 'JQ231114.1': '1.000'}\n"
     ]
    }
   ],
   "source": [
    "print rec_mislabelled"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude the record ids that have been identified as mislabelled from the total list of ids to produce the pool from which random records will be drawn."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1736\n"
     ]
    }
   ],
   "source": [
    "pool = []\n",
    "\n",
    "for i in full_id_list:\n",
    "    if not i in rec_mislabelled.keys():\n",
    "        pool.append(i)\n",
    "        \n",
    "print len(pool)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Bin records by species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "species={}\n",
    "count = 0\n",
    "for r in pj.records:\n",
    "    if r.id in pool:\n",
    "        count+=1\n",
    "        source = [f for f in r.features if f.type == 'source'][0]\n",
    "        if not \" \".join(source.qualifiers['organism'][0].split()[0:2]) in species:\n",
    "            species[\" \".join(source.qualifiers['organism'][0].split()[0:2])] = []\n",
    "        species[\" \".join(source.qualifiers['organism'][0].split()[0:2])].append(r.id)\n",
    "        \n",
    "total=len(species)\n",
    "\n",
    "print \"Total number of records: %i\" %count\n",
    "print \"Number of species (see below for full list): %i\" %len(species)\n",
    "print species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For every species extract at most a total of 5 record ids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "274\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "to_keep = []\n",
    "for sp in species:\n",
    "#    print sp\n",
    "#    print len(species[sp])\n",
    "    if len(species[sp]) > 5:\n",
    "        to_keep.extend(random.sample(species[sp], 5))\n",
    "    else:\n",
    "        to_keep.extend(species[sp])\n",
    "        \n",
    "\n",
    "print len(to_keep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove _Petromyzon marinus_ (Genbank accession: NC_001626.1) from the list of focal records for aesthetic reasons."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "273\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(to_keep)):\n",
    "    if 'NC_001626.1' == to_keep[i]:\n",
    "        del to_keep[i]\n",
    "        break\n",
    "print len(to_keep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add the mislabelled ids to the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "297\n"
     ]
    }
   ],
   "source": [
    "to_keep.extend(rec_mislabelled.keys())\n",
    "print len(to_keep)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we've generated a random list of records, collect the ids of the remaining sequences in a list to be excluded from the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1463\n"
     ]
    }
   ],
   "source": [
    "to_drop = {'CytB': []}\n",
    "\n",
    "for s in full_id_list:\n",
    "    if not s in to_keep:\n",
    "        to_drop['CytB'].append(s)\n",
    "\n",
    "print len(to_drop['CytB'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write out the good records in Genbank format to be analysed with SATIVA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "seqs = []\n",
    "from Bio import SeqIO\n",
    "for r in pj.records:\n",
    "    if r.id in to_keep:\n",
    "        seqs.append(r)\n",
    "\n",
    "out = open('CytB_nr_pre_Sativa.gb','w')\n",
    "SeqIO.write(seqs, out, 'genbank')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Exclude records from Reprophylo data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.exclude(start_from_max=False, **to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure alignment object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mafft --localpair --maxiterate 1000 428121473691046.2_CDS_proteins_CytB.fasta\n"
     ]
    }
   ],
   "source": [
    "mafft_linsi = AlnConf(pj,                              # The Project\n",
    "                    method_name='mafftLinsi',        # Any unique method name,\n",
    "                    CDSAlign=True,                   # Use this method to align\n",
    "                    codontable=2,                    # The genetic code that\n",
    "                    program_name='mafft',            # mafft or muscle.\n",
    "                    cmd='mafft',                     # The command on your machine\n",
    "                    loci=['CytB'],                 # A list of loci names to align.vel.\n",
    "                    cline_args={'localpair': True,# Program specific keywords and arguments.\n",
    "                                'maxiterate': 1000}# cine_args=={} by default, which will\n",
    "                    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.align([mafft_linsi])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CytB@mafftLinsi_aln.fasta']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pj.write_alns(id=['record_id','source_organism'], format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine custom clipping points in alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi\n",
      "\n",
      "Finding forward clip position:\n",
      "\n",
      "1 26 0.700673400673\n",
      "2 27 0.776767676768\n",
      "3 28 0.783838383838\n",
      "4 29 0.790909090909\n",
      "5 30 0.799326599327\n",
      "6 31 0.805050505051\n",
      "7 32 0.810774410774\n",
      "8 33 0.818855218855\n",
      "9 34 0.824915824916\n",
      "10 35 0.830976430976\n",
      "Suggested forward clip at: 25 (10,35)\n",
      "\n",
      "Finding reverse clip position:\n",
      "\n",
      "1 1145 0.701346801347\n",
      "2 1144 0.705723905724\n",
      "3 1143 0.710101010101\n",
      "4 1142 0.714478114478\n",
      "5 1141 0.715488215488\n",
      "6 1140 0.716498316498\n",
      "7 1139 0.717508417508\n",
      "8 1138 0.718518518519\n",
      "9 1137 0.719528619529\n",
      "10 1136 0.720538720539\n",
      "Suggested reverse clip at: 1146 (10,1136)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "per_col=[]\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi')\n",
    "\n",
    "\n",
    "icol=0\n",
    "for a in aln[1,:]:\n",
    "    ACGT_count=0\n",
    "    col=aln[:,icol]\n",
    "    for i in range(len(col)):\n",
    "        if col[i] in ['A','G','C','T']:\n",
    "            ACGT_count+=1\n",
    "#    print pj.fa('CytB@mafftLinsi')[:,a]\n",
    "    per_col.append(float(ACGT_count)/len(col))\n",
    "    icol+=1\n",
    "\n",
    "wsize=10\n",
    "min_prop=0.7\n",
    "    \n",
    "print \"\\nFinding forward clip position:\\n\"\n",
    "forw_count=0\n",
    "forw_clip=0\n",
    "for i in range(len(per_col)-wsize):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        forw_count+=1\n",
    "        forw_clip=i\n",
    "        print forw_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        forw_count=0\n",
    "        \n",
    "    if forw_count >= wsize:\n",
    "        break\n",
    "        \n",
    "print \"Suggested forward clip at: %s (%s,%s)\" %(forw_clip-wsize,forw_count,forw_clip)\n",
    "\n",
    "print \"\\nFinding reverse clip position:\\n\"\n",
    "reve_count=0\n",
    "reve_clip=0\n",
    "for i in reversed(range(len(per_col)-wsize)):\n",
    "        \n",
    "    if np.mean(per_col[i:i+10]) > min_prop:\n",
    "        reve_count+=1\n",
    "        reve_clip=i\n",
    "        print reve_count,i,np.mean(per_col[i:i+wsize])\n",
    "    else:\n",
    "        reve_count=0\n",
    "        \n",
    "    if reve_count >= wsize:\n",
    "        break\n",
    "    \n",
    "print \"Suggested reverse clip at: %s (%s,%s)\" %(reve_clip+wsize,reve_count,reve_clip)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clip Alignment using Biopython."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "returning alignment object CytB@mafftLinsi\n"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aln = pj.fa('CytB@mafftLinsi')\n",
    "\n",
    "clipped_aln = aln[:,forw_clip-wsize:reve_clip+wsize]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the clipped alignemnt (which is a Biopython alignment object) to a ReproPhylo trimmed alignment object, so that we can resume working with it in ReproPhylo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.trimmed_alignments[\"CytB@mafftLinsi_aln_clipped\"] = clipped_aln"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write clipped alignment to file for downstream taxonomic curation using SATIVA.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "out=open('CytB@mafftLinsi_aln_clipped.phy', 'w')\n",
    "AlignIO.write(clipped_aln, out, 'phylip-relaxed')\n",
    "out.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Configure tree inference method - RaxML."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 431761473691128.36_CytB@mafftLinsi_aln_clipped0 -p 553 -s 431761473691128.36_CytB@mafftLinsi_aln_clipped.fasta -T 5 -x 372 -N 100\n"
     ]
    }
   ],
   "source": [
    "raxml = RaxmlConf(pj, method_name='raxml-mafftLinsi-clipped', \n",
    "                 program_name='raxmlHPC-PTHREADS-SSE3',\n",
    "                 keepfiles=True,\n",
    "                 cmd='raxmlHPC-PTHREADS-SSE3',\n",
    "                 preset='fa',\n",
    "                 alns=['CytB@mafftLinsi_aln_clipped'],\n",
    "                 model='GAMMA',\n",
    "                 threads=5,\n",
    "                 cline_args={'-#': 100})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Infer Maximum likelihood tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.tree([raxml])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RaxmlConf named raxml-mafftLinsi-clipped with ID 431761473691128.36\n",
      "Alignments: CytB@mafftLinsi_aln_clipped \n",
      "Created on: Mon Sep 12 14:38:48 2016\n",
      "Commands:\n",
      "CytB@mafftLinsi_aln_clipped:\n",
      "raxmlHPC-PTHREADS-SSE3 -f a -m GTRGAMMA -n 431761473691128.36_CytB@mafftLinsi_aln_clipped0 -p 553 -s 431761473691128.36_CytB@mafftLinsi_aln_clipped.fasta -T 5 -x 372 -N 100\n",
      "\n",
      "Environment:\n",
      "Platform: Linux-3.13.0-54-generic-x86_64-with-Ubuntu-14.04-trusty\n",
      " Processor: x86_64\n",
      " Python build: defaultMar 22 2014 22:59:56\n",
      " Python compiler: GCC 4.8.2\n",
      " Python implementation: CPython\n",
      " Python version: 2.7.6\n",
      " ete2 version: 2.2.1072\n",
      " biopython version: 1.68\n",
      " dendropy version: 3.12.2\n",
      " cloud version: 2.8.5\n",
      " reprophylo version 1.3\n",
      " User: chrishah-office\n",
      " Program and version: raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12\n",
      " Program reference: A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics\n",
      " 2014.\n",
      "execution time:\n",
      "2039.5449481\n",
      "\n",
      "==============================\n",
      "Core Methods section sentence:\n",
      "==============================\n",
      "Phylogenetic trees were reconstructed from the dataset(s) CytB@mafftLinsi_aln_clipped using the program raxmlHPC-PTHREADS-SSE3RAxML version 8.0.12 [1].\n",
      "\n",
      "Reference:\n",
      "A. Stamatakis: RAxML Version 8: A tool for Phylogenetic Analysis and Post-Analysis of Large Phylogenies. In Bioinformatics, 2014.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print pj.used_methods['raxml-mafftLinsi-clipped']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'reduced_tree.pkl'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_pj(pj, 'reduced_tree.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's visualize the tree.\n",
    "\n",
    "Prepare tree annotation:\n",
    " - define outgroups\n",
    " - specify support coloration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.add_qualifier_from_source('organism')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for species in ['Lampetra']:\n",
    "    pj.if_this_then_that(species, 'organism', 'outgroup', 'outgroup',\n",
    "                         mode='part')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "supports = {'black': [100,95],\n",
    "            'dimgray': [95,75],\n",
    "            'silver': [75,50]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=50, scale=500,\n",
    "            html='mafftLinsi-clipped-raxml.html')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a static image of the final tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png CytB_mafftLinsi_clipped_raxml_dirty_multifurc50.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternative formatting of tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=None, scale=500,\n",
    "            html='mafftLinsi-clipped-raxml.html')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png CytB_mafftLinsi_clipped_raxml_dirty_multifurcNone.png"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We move to running SATIVA at this stage to identify putatively mislabelled sequencing records.\n",
    "\n",
    "\n",
    "##============================================##\n",
    "\n",
    "__Run SATIVA__ see notebook [here]().\n",
    "\n",
    "##============================================##\n",
    "\n",
    "\n",
    "After we have run SATIVA on the alignment that we produced above.\n",
    "\n",
    "Now we are going to highlight the records that SATIVA considers mislabelled in the tree. We are also going to add a heatmap to the tree illustrating the confidence that SATIVA has in a given call.\n",
    "Parse SATIVA output and record putatively mislabelled sequence records and the confidence of the call.\n",
    "\n",
    "Annotate tree with SATIVA results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "rec_mislabelled = {}\n",
    "sativa=open('../run_SATIVA/CytB.mis', 'r')\n",
    "\n",
    "for line in sativa:\n",
    "    if not line.startswith(\";\"):\n",
    "        rec_mislabelled[line.split(\"\\t\")[0]] = line.split(\"\\t\")[4]\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add two new qualifiers containing info about he SATIVA results to the sequence features."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for r in pj.records:\n",
    "    if r.id in rec_mislabelled:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['yes']\n",
    "            f.qualifiers['mislabeled_conf'] = [rec_mislabelled[r.id]]\n",
    "    else:\n",
    "        for f in r.features:\n",
    "            f.qualifiers['mislabeled'] = ['no']\n",
    "            f.qualifiers['mislabeled_conf'] = [0.0001]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.propagate_metadata()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare tree annotation:\n",
    "\n",
    " - color records that were identified by SATIVA as being mislabelled.\n",
    " - add heatmap with mislabel confidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "leaf_colors = {'yes':'red', 'no': 'black'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pj.clear_tree_annotations()\n",
    "pj.annotate('.', 'outgroup', 'outgroup', \n",
    "            ['source_organism', 'record_id'], \n",
    "            node_support_dict=supports, multifurc=10,\n",
    "            html='mafftLinsi-clipped-raxml-SATIVA.html',\n",
    "            fsize=13, scale=500,\n",
    "            heat_map_meta = ['mislabeled_conf'], heat_map_colour_scheme = 1,\n",
    "            leaf_node_color_meta = 'mislabeled', leaf_label_colors = leaf_colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create static image of final annoated tree."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "last_png=$(ls -hlrt *CytB@mafftLinsi_aln_clipped.png | tail -n 1 | perl -ne 'chomp; @a=split(\" \"); print \"$a[-1]\"')\n",
    "cp $last_png CytB_mafftLinsi_clipped_raxml_SATIVA_annotated.png"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
